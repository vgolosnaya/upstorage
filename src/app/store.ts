import { configureStore } from '@reduxjs/toolkit';
import filesReducer from './file/files.reducer';

export const store = configureStore({
    reducer: {
        files: filesReducer,
    },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
