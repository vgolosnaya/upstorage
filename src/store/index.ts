import { combineReducers, configureStore } from '@reduxjs/toolkit';
import filesReducer from 'store/files/files.reducer';
import { IFile } from 'services/FileReader.service';
import filesMiddleware from 'store/files/files.middleware';
import StorageService from 'services/Storage.service';
import errorsReducer, { ErrorState } from 'store/errors/errors.reducer';

const storageService = new StorageService<IFile>('files');

export const store = configureStore({
    reducer: combineReducers({
        files: filesReducer,
        errors: errorsReducer,
    }),
    middleware: [ filesMiddleware(storageService) ]
});

export type AppDispatch = typeof store.dispatch;
export type RootState = {
    files: IFile[],
    errors: ErrorState[],
};
