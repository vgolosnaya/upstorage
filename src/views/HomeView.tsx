import React, { useContext } from 'react';
import { styled } from '@mui/material/styles';
import {
    Button,
    Container,
    Paper,
    Table,
    TableBody,
    TableCell,
    TableContainer,
    TableHead,
    TableRow, Typography
} from '@mui/material';
import LABELS from '../labels';
import { Event } from '../types';
import filesActions from '../app/file/file.actions';
import { useAppDispatch } from '../app/hooks';
import { FileReaderContext } from '../index';


function createData(
    fileName: string,
    fileExtension: 'png' | 'svg',
    dateTime: string,
) {
    return { fileName, fileExtension, dateTime };
}

const rows = [
    createData('test_png1', 'png', 'test'),
    createData('test_svg1', 'svg', 'test'),
    createData('test_svg2', 'svg', 'test'),
    createData('test_png2', 'png', 'test')
];

const HomeView = (): JSX.Element => {
    const Input = styled('input')({
        display: 'none',
    });
    const fileReader = useContext(FileReaderContext);
    const dispatch = useAppDispatch();
    
    const onFileUpload = async(event: Event<HTMLInputElement>) => {
        // eslint-disable-next-line no-unsafe-optional-chaining
        const files = Array.from(event?.target.files || []);
       
        while(files.length) {
            const file = files.shift()!;
            dispatch(filesActions.readFileSuccess(await fileReader.readFile(file)));
        }
    };
    
    return (
        <Container maxWidth="xl">
            <label htmlFor="contained-button-file">
                <Input id="contained-button-file" multiple type="file" onChange={onFileUpload}/>
                <Button variant="contained" component="span">
                    {LABELS.upload}
                </Button>
            </label>
            <TableContainer component={Paper}>
                <Typography variant="overline" component="div" gutterBottom>
                    {LABELS.recently_loaded}
                </Typography>
                <Table sx={{ minWidth: 650 }} aria-label="simple table">
                    <TableHead>
                        <TableRow>
                            <TableCell>FileName</TableCell>
                            <TableCell align="right">Extension</TableCell>
                            <TableCell align="right">Datetime</TableCell>
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        {rows.map((row) => (
                            <TableRow
                                key={row.fileName}
                                sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
                            >
                                <TableCell component="th" scope="row">
                                    {row.fileName}
                                </TableCell>
                                <TableCell align="right">{row.fileExtension}</TableCell>
                                <TableCell align="right">{row.dateTime}</TableCell>
                            </TableRow>
                        ))}
                    </TableBody>
                </Table>
            </TableContainer>
        </Container>
    );
};

export default HomeView;
