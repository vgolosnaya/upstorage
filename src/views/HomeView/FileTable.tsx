import { Paper, Table, TableBody, TableCell, TableContainer, TableHead, TableRow, Typography } from '@mui/material';
import LABELS from 'labels';
import React from 'react';
import { useAppSelector } from 'store/hooks';
import filesSelectors from 'store/files/files.selectors';

const FileTable = ():JSX.Element => {
    const files = useAppSelector(filesSelectors.getRecent);
    return (
        <TableContainer component={Paper}>
            <Typography variant="overline" component="div" gutterBottom>
                {LABELS.recently_loaded}
            </Typography>
            <Table sx={{ minWidth: 650 }} aria-label="simple table">
                <TableHead>
                    <TableRow>
                        <TableCell>FileName</TableCell>
                        <TableCell align="right">Extension</TableCell>
                        <TableCell align="right">Datetime</TableCell>
                    </TableRow>
                </TableHead>
                <TableBody>
                    {files.map((file) => (
                        <TableRow
                            key={file.id}
                            sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
                        >
                            <TableCell component="th" scope="row">
                                {file.fileName}
                            </TableCell>
                            <TableCell align="right">{file.type}</TableCell>
                            <TableCell align="right">{file.relativeDateTime}</TableCell>
                        </TableRow>
                    ))}
                </TableBody>
            </Table>
        </TableContainer>
    );
};

export default FileTable;