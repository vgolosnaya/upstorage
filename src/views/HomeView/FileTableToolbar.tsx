import React, { useContext } from 'react';
import { styled } from '@mui/material/styles';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';
import Grid from '@mui/material/Grid';
import { Event } from 'types';
import { useAppDispatch } from 'store/hooks';
import { fileActionCreators } from 'store/files/files.actions';
import { errorActionCreators } from 'store/errors/errors.actions';
import FileReaderContext from 'FileReaderContext';
import { FileException } from 'config';
import LABELS from 'labels';

const FileTableToolbar = (): JSX.Element => {
    const Input = styled('input')({
        display: 'none',
    });
    const fileReader = useContext(FileReaderContext);
    const dispatch = useAppDispatch();
    
    const onFileUpload = async(event: Event) => {
        const target = event?.target as HTMLInputElement;
        const files = Array.from(target.files || []);
        
        while(files.length) {
            const file = files.shift()!;
            try {
                const result = await fileReader.readFile(file);
                dispatch(fileActionCreators.saveFileToStorage(result));
            } catch(error) {
                console.log(error);
                if(error instanceof FileException) {
                    dispatch(errorActionCreators.showError(error.message));
                }
            }
        }
    };
    
    return (
        <Grid container rowSpacing={1}>
            <Grid item xs={6}>
                <Typography variant="h4" component="div" gutterBottom>
                    {LABELS.recently_loaded}
                </Typography>
            </Grid>
            <Grid item xs={6} justifyContent="end" display="flex">
                <label htmlFor="contained-button-file">
                    <Input accept="text/csv,image/png" id="contained-button-file" multiple type="file"
                        onInput={onFileUpload} data-testid="upload"/>
                    <Button variant="contained" component="span">
                        {LABELS.upload}
                    </Button>
                </label>
            </Grid>
        </Grid>
    );
};

export default FileTableToolbar;
