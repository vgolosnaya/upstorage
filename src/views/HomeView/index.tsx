import React, { useContext } from 'react';
import { styled } from '@mui/material/styles';
import { Button } from '@mui/material';
import LABELS from 'labels';
import { Event } from 'types';
import { fileActionCreators } from 'store/files/files.actions';
import { useAppDispatch } from 'store/hooks';
import { FileReaderContext } from 'index';
import FileTable from './FileTable';

const HomeView = (): JSX.Element => {
    const Input = styled('input')({
        display: 'none',
    });
    const fileReader = useContext(FileReaderContext);
    const dispatch = useAppDispatch();
    
    const onFileUpload = async(event: Event<HTMLInputElement>) => {
        const files = Array.from(event?.target.files || []);
        
        while(files.length) {
            const file = files.shift()!;
            const result = await fileReader.readFile(file);
            dispatch(fileActionCreators.saveFileToStorage(result));
        }
    };
    
    return (
        <div className="container">
            <label htmlFor="contained-button-file">
                <Input accept="text/csv,image/png" id="contained-button-file" multiple type="file" onChange={onFileUpload}/>
                <Button variant="contained" component="span">
                    {LABELS.upload}
                </Button>
            </label>
            <FileTable/>
        </div>
    );
};

export default HomeView;
