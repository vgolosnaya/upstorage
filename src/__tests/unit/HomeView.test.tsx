import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from 'store';
import React from 'react';
import HomeView from 'views/HomeView';

describe('[SNAPSHOT] Home page:', ()=>{
    test('Should match snapshot', () => {
        const component = render(
            <Provider store={store}>
                <HomeView/>
            </Provider>
        );
        
        expect(component.baseElement).toMatchSnapshot();
    });
});