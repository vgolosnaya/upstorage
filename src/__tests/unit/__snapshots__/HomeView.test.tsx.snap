// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SNAPSHOT] Home page: Should match snapshot 1`] = `
<body>
  <div>
    <div
      class="container"
      data-testid="homepage"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-1sjizby-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1vw6mcs-MuiTypography-root"
          >
            Recently loaded
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-jlvj17-MuiGrid-root"
        >
          <label
            for="contained-button-file"
          >
            <input
              accept="text/csv,image/png"
              class="css-1hyfx7x"
              data-testid="upload"
              id="contained-button-file"
              multiple=""
              type="file"
            />
            <span
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
              role="button"
              tabindex="0"
            >
              UPLOAD
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
          </label>
        </div>
      </div>
      <span
        class="MuiTypography-root MuiTypography-overline css-b3fch5-MuiTypography-root"
      >
        You don't have any file loaded
      </span>
    </div>
  </div>
</body>
`;
